### ISSUE TITLE
:Termdebug Fails to Follow Code Execution (Cursor/Highlighted Line Does Not Move)

### ISSUE THREAD

**Issue by NoSuck** [vim/vim#2417]
Hello.  Termdebug is a boon to programmers who use Vim.  Thank you for introducing a feature I did not even realize I needed.

Perhaps related to #2154, GDB execution does not seem to follow (i.e. automatically highlight) source code in the originating Vim buffer.  Unlike #2154, changing terminal dimensions does not seem to affect the issue for me.  The attached [demonstration](https://nosuckdotorg.files.wordpress.com/2017/12/2017-12-05-termdebug-does-not-follow-cursor-8-0-1297-4-13-12-1-arch-x86_64-gnu-linux.gif) was taken from 8.0.1297, 4.13.12-1-ARCH (x86_64 GNU/Linux).  $TERM is "xterm".  Can anyone provide pointers or troubleshooting steps?

![2017-12-05 termdebug does not follow cursor 8 0 1297 4 13 12-1-arch x86_64 gnu-linux](https://user-images.githubusercontent.com/10365452/33637840-2fbf38ce-d9f1-11e7-9e2d-1be7d374ccdb.gif)

**Comment by NoSuck** [comment]
Perhaps I should have mentioned this explicitly:  In the demonstration, I use both GDB "next" commands and mouse interaction with the Termdebug toolbar (upper-left corner).  Neither results in Vim highlighting GDB's line of execution (and the latter produces "Switching to thread" messages).

GDB's version is 8.0.1.

**Comment by brammool** [comment]
> Hello.  Termdebug is a boon to programmers who use Vim.  Thank you for introducing a feature I did not even realize I needed.
> 
> Perhaps related to #2154, GDB execution does not seem to follow (i.e. automatically highlight) source code in the originating Vim buffer.  Unlike #2154, changing terminal dimensions does not seem to affect the issue for me.  The attached [demonstration](https://nosuckdotorg.files.wordpress.com/2017/12/2017-12-05-termdebug-does-not-follow-cursor-8-0-1297-4-13-12-1-arch-x86_64-gnu-linux.gif) was taken from 8.0.1297, 4.13.12-1-ARCH (x86_64 GNU/Linux).  $TERM is "xterm".  Can anyone provide pointers?

Have you tried the example session?  :help termdebug-example

It might be related to multi-threading.  If you open the communication
window, perhaps you can see what gdb sends when using a "next" command.

-- 
From "know your smileys":
 y:-)	Bad toupee

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by NoSuck** [comment]
I apologize for disappearing.  I am back to C programming (this time on a different system) and tried out vim's `:Termdebug` only to again run into this issue.

Thank you for the tip regarding `:help termdebug-example`.  I followed it to a T with the exception of appending `-u NONE` to vim calls.  However, I do suspect the validity of this example.  Why is `vim` and not `./vim` debugged, for instance?

![Debugging Termdebug](https://user-images.githubusercontent.com/10365452/67906895-732e2100-fb4c-11e9-97d0-b201deed6c22.gif)

That said, I also tried personal projects--but behavior remains consistent:  Vim does not even open code files, much less follow along with gdb.  I have tried adding locations with gdb `dir` commands to no avail.

I did notice that setting a breakpoint at `malloc()` resulted in vim opening `malloc.c` to the correct line.  Here is output from the gdb communication window for both `malloc` and `ex_help`.  I see that the latter is missing `file`, `fullname`, and `line`.  Any idea why?

Making a breakpoint and breaking at `malloc()`:
```
(gdb)
=breakpoint-created,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x00012634",at="<malloc@plt>",thread-groups=["i1"],times="0",original-location="malloc"}
=thread-group-started,id="i1",pid="4019"
=thread-created,id="1",group-id="i1"
=library-loaded,id="/lib/ld-linux-armhf.so.3",target-name="/lib/ld-linux-armhf.so.3",host-name="/lib/ld-linux-armhf.so.3",symbols-loaded="0",thread-group="i1",ranges=[{from="0x76fc1a30",to="0x76fdd150"}]
=breakpoint-modified,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x76fd8f2c",func="malloc",file="dl-minimal.c",fullname="/usr/src/glibc/glibc-2.28/elf/dl-minimal.c",line="50",thread-groups=["i1"],times="0",original-location="malloc"}
^running
*running,thread-id="all"
=breakpoint-modified,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x76fd8f2c",func="malloc",file="dl-minimal.c",fullname="/usr/src/glibc/glibc-2.28/elf/dl-minimal.c",line="50",thread-groups=["i1"],times="1",original-location="malloc"}
~"\n"
~"Breakpoint 1, malloc (n=n@entry=753) at dl-minimal.c:50\n"
~"50\t  if (alloc_end == 0)\n"
*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x76fd8f2c",func="malloc",args=[{name="n",value="753"},{name="n@entry",value="753"}],file="dl-minimal.c",fullname="/usr/src/glibc/glibc-2.28/elf/dl-minimal.c",line="50"},thread-id="1",stopped-threads="all",core="2"
```

Making a breakpoint at `ex_help()`:
```
(gdb)
4=breakpoint-created,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x00071194",at="<ex_help+8>",thread-groups=["i1"],times="0",original-location="ex_help"}
```

Breaking at `ex_help()`:
```
=breakpoint-modified,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x00071194",at="<ex_help+8>",thread-groups=["i1"],times="1",original-location="ex_help"}
~"\n"
~"Breakpoint 1, 0x00071194 in ex_help ()\n"
*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x00071194",func="ex_help",args=[]},thread-id="1",stopped-threads="all",core="3"
```

**Comment by brammool** [comment]
The message from gdb about where it stopped does not included the file location.  You have to make sure that the executable is not stripped.

**Comment by NoSuck** [comment]
Perhaps vim's example should be updated, in that case.

Here is the gdb communication window's output from a personal project.  Should *this* open files in vim's source window?  On my system, it does not. 

```
*running,thread-id="4"
=breakpoint-modified,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x0002ae9c",func="WsRealloc",file
="Module/Locale.c",fullname="/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\343\202\260\343
\203\251\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c",line="174",thread-groups=["i1"],times
="1",original-location="WsRealloc"}
~"\n"
~"Thread 1 \"nsmt\" hit Breakpoint 1, WsRealloc (ws=0x1b995c8 L\"\347\267\250\351\233\206\345\261\245\346\255\264\", sF
ormat=0x4b3d8 \"%u\360\237\206\232%u\") at Module/Locale.c:174\n"
~"174\t\tva_start (args, sFormat);\n"
*stopped,reason="breakpoint-hit",disp="keep",bkptno="1",frame={addr="0x0002ae9c",func="WsRealloc",args=[{name="ws",valu
e="0x1b995c8 L\"\347\267\250\351\233\206\345\261\245\346\255\264\""},{name="sFormat",value="0x4b3d8 \"%u\360\237\206\23
2%u\""}],file="Module/Locale.c",fullname="/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\34
3\202\260\343\203\251\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c",line="174"},thread-id="1
",stopped-threads="all",core="1"
*running,thread-id="all"
*stopped,reason="end-stepping-range",frame={addr="0x0002aea4",func="WsRealloc",args=[{name="ws",value="0x1b995c8 L\"\34
7\267\250\351\233\206\345\261\245\346\255\264\""},{name="sFormat",value="0x4b3d8 \"%u\360\237\206\232%u\""}],file="Modu
le/Locale.c",fullname="/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\343\202\260\343\203\2
51\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c",line="179"},thread-id="1",stopped-threads="
all",core="1"
```

Output of gdb's `dir`:
```
Source directories searched: /home/userx/俺/開発/プログラム/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Header:/home/
userx/俺/開発/プログラム/2014-06-25 No Suck MOD Tracker/2016-09-19 C:/home/userx/俺/開発/プログラム/2014-06-25 No Suc
k MOD Tracker/2016-09-19 C/Module:/usr/src/glibc/current/wctype:/usr/src/glibc/current/wcsmbs:/usr/src/glibc/current/
timezone:/usr/src/glibc/current/time:/usr/src/glibc/current/termios:/usr/src/glibc/current/sysvipc:/usr/src/glibc/cur
rent/sysdeps:/usr/src/glibc/current/support:/usr/src/glibc/current/sunrpc:/usr/src/glibc/current/string:/usr/src/glib
c/current/streams:/usr/src/glibc/current/stdlib:/usr/src/glibc/current/stdio-common:/usr/src/glibc/current/soft-fp:/u
sr/src/glibc/current/socket:/usr/src/glibc/current/signal:/usr/src/glibc/current/shadow:/usr/src/glibc/current/setjmp
:/usr/src/glibc/current/rt:/usr/src/glibc/current/resource:/usr/src/glibc/current/resolv:/usr/src/glibc/current/pwd:/
usr/src/glibc/current/posix:/usr/src/glibc/current/nss:/usr/src/glibc/current/nscd:/usr/src/glibc/current/nptl_db:/us
r/src/glibc/current/nptl:/usr/src/glibc/current/nis:/usr/src/glibc/current/misc:/usr/src/glibc/current/math:/usr/src/
glibc/current/malloc:/usr/src/glibc/current/mach:/usr/src/glibc/current/login:/usr/src/glibc/current/localedata:/usr/
src/glibc/current/locale:/usr/src/glibc/current/libio:/usr/src/glibc/current/io:/usr/src/glibc/current/intl:/usr/src/
glibc/current/inet:/usr/src/glibc/current/include:/usr/src/glibc/current/iconvdata:/usr/src/glibc/current/iconv:/usr/
src/glibc/current/hurd:/usr/src/glibc/current/htl:/usr/src/glibc/current/hesiod:/usr/src/glibc/current/gshadow:/usr/s
rc/glibc/current/grp:/usr/src/glibc/current/gnulib:/usr/src/glibc/current/gmon:/usr/src/glibc/current/fbtl_db:/usr/sr
c/glibc/current/fbtl:/usr/src/glibc/current/elf:/usr/src/glibc/current/dlfcn:/usr/src/glibc/current/dirent:/usr/src/g
libc/current/debug:/usr/src/glibc/current/ctype:/usr/src/glibc/current/csu:/usr/src/glibc/current/crypt:/usr/src/glib
c/current/catgets:/usr/src/glibc/current/bits:/usr/src/glibc/current/assert:/usr/src/glibc/current/argp:$cdir:$cwd   
```

```
■ readelf -s ./nsmt | grep WsRealloc
  2828: 0002ae74   288 FUNC    GLOBAL DEFAULT   13 WsRealloc
```

```VIM - Vi IMproved 8.1 (2018 May 18, compiled Jun 15 2019 16:41:15)
Included patches: 1-875, 878, 884, 948, 1046, 1365-1368, 1382, 1401
Modified by team+vim@tracker.debian.org
Compiled by team+vim@tracker.debian.org
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     +tag_old_static
+arabic            +farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    +tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
+balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
+browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +textprop
+byte_offset       -hangul_input      +num64             +timers
+channel           +iconv             +packages          +title
+cindent           +insert_expand     +path_extra        +toolbar
+clientserver      +job               +perl              +user_commands
+clipboard         +jumplist          +persistent_undo   +vartabs
+cmdline_compl     +keymap            +postscript        +vertsplit
+cmdline_hist      +lambda            +printer           +virtualedit
+cmdline_info      +langmap           +profile           +visual
+comments          +libcall           -python            +visualextra
+conceal           +linebreak         +python3           +viminfo
+cryptv            +lispindent        +quickfix          +vreplace
+cscope            +listcmds          +reltime           +wildignore
+cursorbind        +localmap          +rightleft         +wildmenu
+cursorshape       +lua               +ruby              +windows
+dialog_con_gui    +menu              +scrollbind        +writebackup
+diff              +mksession         +signs             +X11
+digraphs          +modify_fname      +smartindent       -xfontset
+dnd               +mouse             +startuptime       +xim
-ebcdic            +mouseshape        +statusline        +xpm
+emacs_tags        +mouse_dec         -sun_workshop      +xsmp_interact
+eval              +mouse_gpm         +syntax            +xterm_clipboard
+ex_extra          -mouse_jsbterm     +tag_binary        -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/arm-linux-gnueabihf/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/uuid -I/usr/include/glib-2.0 -I/usr/lib/arm-linux-gnueabihf/glib-2.0/include -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-jkgW07/vim-8.1.0875=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/arm-linux-gnueabihf/perl/5.28/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.7/config-3.7m-arm-linux-gnueabihf -lpython3.7m -lcrypt -lpthread -ldl -lutil -lm -L/usr/lib/arm-linux-gnueabihf -ltcl8.6 -ldl -lz -lpthread -lm -lruby-2.5 -lpthread -lgmp -ldl -lcrypt -lm     
```

**Comment by brammool** [comment]
If you follow the example it works, since "make" in the source directory doesn't strip the executable.  But perhaps your CFLAGS doesn't include -g, I'll add a remark about that.

**Comment by NoSuck** [comment]
Thank you.

What about the above?

**Comment by GitMensch** [comment]
> `fullname="/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\343\202\260\343\203\251\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c"`

Whoa, what's that for a filename? Looks like octal encoded character values coming from GDB. (@NoSuck: is that UTF8 or different encoded?)
I see that `s:DecodeMessage()` won't handle that and use it just without the leading backslash, which is definitely wrong and possibly would also be returned in other places.

#9228 provides a starter with TODOs for handling those escape sequences.

@brammool: As I have neither a clue how to decode the octal value strings in viml an if those should be `"\U"` strings nor how vim handles non-ascii filenames (would have to be handled in the calling `s:GetFullname()` it would be nice if you could check fixing `s:DecodeMessage()` and possibly also `s:GetFullname()` or could ping someone who knows about this. Hopefully the referenced PR and this analysis helps a bit.

**Comment by brammool** [comment]
To decode octal you can create a number starting with 0o and evaluate it.  Or just use it as a string.  Not sure what happens with utf8, or what should happen with it.  Perhaps this is a file name with multi-byte characters?

**Comment by jottkaerr** [comment]
> > `fullname="/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\343\202\260\343\203\251\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c"`
> 
> Whoa, what's that for a filename? Looks like octal encoded character values coming from GDB. (@NoSuck: is that UTF8 or different encoded?)

Just execute
```
:echo "/home/userx/\344\277\272/\351\226\213\347\231\272/\343\203\227\343\203\255\343\202\260\343\203\251\343\203\240/2014-06-25 No Suck MOD Tracker/2016-09-19 C/Module/Locale.c"
```

**Comment by lacygoill** [comment]
> it would be nice if you could check fixing s:DecodeMessage()

What's the purpose of `s:DecodeMessage()`?  Looking at the current code (including the PR), it seems that it does this:

   1. make sure that the string starts with a double quote
   2. remove the opening quote
   3. remove anything after the closing quote – if any – including the quote
   4. remove any newline encoded with the escape sequence `\n`
   5. replace any tab character encoded with the escape sequence `\t` with a literal tab character
   6. replace any character encoded with an hexadecimal number, via an escape sequence such as `\x12`, with the actual character
   7. replace any character encoded with an octal number, via an escape sequence such as `\123`, with the actual character

`1.` is already done correctly:

    if a:quotedText[0] != '"'
      echoerr 'DecodeMessage(): missing quote in ' . a:quotedText
      return
    endif

For `2.`:

    let result = a:quotedText->substitute('^"', '', '')

For `2.` and `3.`:

    let result = a:quotedText->substitute('^"\|".*', '', 'g')

For `4.`:

    let result = a:quotedText->substitute('\\n', "\n", 'g')

For `5.`:

    let result = a:quotedText->substitute('\\t', "\t", 'g')

For `6.`:

    let result = a:quotedText->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')

For `7.`:

    let result = a:quotedText->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')

For `2.` to `7.`:

    let result = a:quotedText
        \->substitute('^"\|".*', '', 'g')
        \->substitute('\\n', "\n", 'g')
        \->substitute('\\t', "\t", 'g')
        \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
        \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')

---

Test to illustrate that the code works:
```vim
function DecodeMessage(quotedText)
  return a:quotedText
    \->substitute('^"\|".*', '', 'g')
    \->substitute('\\n', "\n", 'g')
    \->substitute('\\t', "\t", 'g')
    \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
    \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
endfunction
echo DecodeMessage('"aaa\nbbb\t\x63\x63\x63\144\144\144"irrelevant')
```
Outputs:

    aaa
    bbb     cccddd

Notice that:

   - the opening quote was removed
   - everything after the closing quote was removed (i.e. `"irrelevant`)
   - `\n` was correctly translated into a literal newline character
   - `\t` was correctly translated into a literal tab character
   - `\x63\x63\x63` was correctly replaced with `ccc`
   - `\144\144\144` was correctly replaced with `ddd`

---

Note that you could translate everything with a simple `eval()` (after removing what comes after the closing quote):

    echo '"aaa\nbbb\t\x63\x63\x63\144\144\144"irrelevant'
      \->substitute('"[^"]*"\zs.*', '', '')
      \->eval()

    aaa
    bbb     cccddd

But I guess evaluating what could be arbitrary data would make the code too insecure and unpredictable.

---

In any case, I don't see why there is a need for the current long and convoluted while loop.

---

Suggested patch (untested):
```diff
diff --git a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
index 7087f8b25..076d4e02b 100644
--- a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
+++ b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
@@ -566,26 +566,12 @@ func s:DecodeMessage(quotedText)
     echoerr 'DecodeMessage(): missing quote in ' . a:quotedText
     return
   endif
-  let result = ''
-  let i = 1
-  while a:quotedText[i] != '"' && i < len(a:quotedText)
-    if a:quotedText[i] == '\'
-      let i += 1
-      if a:quotedText[i] == 'n'
-	" drop \n
-	let i += 1
-	continue
-      elseif a:quotedText[i] == 't'
-	" append \t
-	let i += 1
-	let result .= "\t"
-	continue
-      endif
-    endif
-    let result .= a:quotedText[i]
-    let i += 1
-  endwhile
-  return result
+  return a:quotedText
+    \->substitute('^"\|".*', '', 'g')
+    \->substitute('\\n', "\n", 'g')
+    \->substitute('\\t', "\t", 'g')
+    \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
+    \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
 endfunc
 
 " Extract the "name" value from a gdb message with fullname="name".
```

**Comment by lacygoill** [comment]
Ah, you want to drop the newlines, not translate them.  New patch (still untested):
```diff
diff --git a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
index 7087f8b25..076d4e02b 100644
--- a/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
+++ b/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
@@ -566,26 +566,12 @@ func s:DecodeMessage(quotedText)
     echoerr 'DecodeMessage(): missing quote in ' . a:quotedText
     return
   endif
-  let result = ''
-  let i = 1
-  while a:quotedText[i] != '"' && i < len(a:quotedText)
-    if a:quotedText[i] == '\'
-      let i += 1
-      if a:quotedText[i] == 'n'
-	" drop \n
-	let i += 1
-	continue
-      elseif a:quotedText[i] == 't'
-	" append \t
-	let i += 1
-	let result .= "\t"
-	continue
-      endif
-    endif
-    let result .= a:quotedText[i]
-    let i += 1
-  endwhile
-  return result
+  return a:quotedText
+    \->substitute('^"\|".*', '', 'g')
+    \->substitute('\\n', '', 'g')
+    \->substitute('\\t', "\t", 'g')
+    \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
+    \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
 endfunc
 
 " Extract the "name" value from a gdb message with fullname="name".
```

**Comment by GitMensch** [comment]
Thank you for inspecting that - I think this should go as a patch. Note: instead of substituting the leading quote away - which is checked to exist beforehand - just using a sub-string would likely be reasonable.
My main question for that would be the performance - that function may be executed very often and the loop _may_ be much faster than substitute an eval (but that part is likely to come in any case, see #9228).

**Comment by lacygoill** [comment]
> Note: instead of substituting the leading quote away - which is checked to exist beforehand - just using a sub-string would likely be reasonable.

I thought about that, and yes it's possible that it would be faster.  I preferred the `substitute()` way because it seemed more readable, but that's subjective.

   > My main question for that would be the performance - that function may be executed very often and the loop may be much faster than substitute an eval (but that part is likely to come in any case, see #9228).

It *seems* that my suggestion is already at least 10 times faster than the current loop (while doing more character translations):

    vim -Nu NONE -S <(cat <<'EOF'
        function DecodeMessage(quotedText)
          let result = ''
          let i = 1
          while a:quotedText[i] != '"' && i < len(a:quotedText)
            if a:quotedText[i] == '\'
              let i += 1
              if a:quotedText[i] == 'n'
                " drop \n
                let i += 1
                continue
              elseif a:quotedText[i] == 't'
                " append \t
                let i += 1
                let result ..= "\t"
                continue
              endif
            endif
            let result ..= a:quotedText[i]
            let i += 1
          endwhile
          return result
        endfunction
        let time = reltime()
        const TIMES = 10000
        for i in TIMES->range()
            eval DecodeMessage('"aaa\nbbb\t\x63\x63\x63\144\144\144"irrelevant')
        endfor
        eval (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run DecodeMessage() ' .. TIMES .. ' times')->setline(1)
    EOF
    )

    6.048 seconds to run DecodeMessage() 10000 times

    vim -Nu NONE -S <(cat <<'EOF'
        function DecodeMessage(quotedText)
          return a:quotedText
            \->substitute('^"\|".*', '', 'g')
            \->substitute('\\n', "\n", 'g')
            \->substitute('\\t', "\t", 'g')
            \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
            \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
        endfunction
        let time = reltime()
        const TIMES = 10000
        for i in TIMES->range()
            eval DecodeMessage('"aaa\nbbb\t\x63\x63\x63\144\144\144"irrelevant')
        endfor
        eval (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run DecodeMessage() ' .. TIMES .. ' times')->setline(1)
    EOF
    )

    0.562 seconds to run DecodeMessage() 10000 times

For better performance, we could also rewrite the function in Vim9:

    def DecodeMessage(quotedText: string): string
    # Decode a message from gdb.  quotedText starts with a ", return the text up
    # to the next ", unescaping characters.
      if quotedText[0] != '"'
        echoerr 'DecodeMessage(): missing quote in ' .. quotedText
        return ''
      endif
      return quotedText
        ->substitute('^"\|".*\|\\n', '', 'g')
        ->substitute('\\t', "\t", 'g')
        ->substitute('\\x\x\x', (m: list<string>): string => eval('"' .. m[0] .. '"'), 'g')
        ->substitute('\\\o\o\o', (m: list<string>): string => eval('"' .. m[0] .. '"'), 'g')
    enddef

or even the [whole plugin](https://github.com/lacygoill/config/blob/master/.vim/pack/mine/opt/termdebug/plugin/termdebug.vim).

**Comment by GitMensch** [comment]
Thanks for the performance checks and notes! So the substitute should likely go in @brammool said he wanted to adjust this and as your version seems preferable to my one I hope that s the one that will land.
I wasn't aware of the functional interface with substitute inline, so also thank you for showing that.

I'm not sure how stable Vim9 is (there is likely an issue or something with updates to that?) and if it would also work with older versions of Vim where the current one works - if yes then it would likely be a good idea (even when this means I have to re-learn much of what I've learned about vim in the last 5 weeks).

**Comment by GitMensch** [comment]
The following function works _nearly_ correct:

```viml
" Decode a message from gdb.  quotedText starts with a ", return the text up
" to the next ", unescaping characters.
func s:DecodeMessage(quotedText)
  if a:quotedText[0] != '"'
    echoerr 'DecodeMessage(): missing quote in ' . a:quotedText
    return
  endif
  return a:quotedText
        \->substitute('^"\|".*', '', 'g')
        \->substitute('\\n', '', 'g')
        \->substitute('\\t', "\t", 'g')
        \->substitute('\\x\x\x', {-> eval('"' .. submatch(0) .. '"')}, 'g')
        \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
        \->substitute('\\\\', '\', 'g')
  return result
endfunc
```

* changed ` \n` to be removed
* replaced `\\` to `\`
* missing: the GDB format for hex is not `\xXX` but `\0xXX`.

Definitely works faster within Termdebug sessions!

**Comment by lacygoill** [comment]
> The following function works nearly correct:

Nice.  You might optimize the code a little more by merging the first 2 substitutes (reducing the number of `substitute()` invocations), since they have the same replacement text (an empty string), and the same flag (`g`):

    " before
    \->substitute('^"\|".*', '', 'g')
    \->substitute('\\n', '', 'g')
    \...

    " after
    \->substitute('^"\|".*\|\\n', '', 'g')
    \...

Also, you can drop the last `:return result`, since you no longer create a return variable, and the previous `:return` makes this line of code unreachable anyway.

**Comment by GitMensch** [comment]
Totally, version with also correct GDB hex replacement:

```viml
" Decode a message from gdb.  quotedText starts with a ", return the text up
" to the next ", unescaping characters (also hex \0xE4 and octal \123), removing newlines.
func s:DecodeMessage(quotedText)
  if a:quotedText[0] != '"'
    echoerr 'DecodeMessage(): missing quote in ' . a:quotedText
    return
  endif
  return a:quotedText
        \->substitute('^"\|".*\|\\n', '', 'g')
        \->substitute('\\t', "\t", 'g')
        \->substitute('\\0x\(\x\x\)', {-> eval('"\x' .. submatch(1) .. '"')}, 'g')
        \->substitute('\\\o\o\o', {-> eval('"' .. submatch(0) .. '"')}, 'g')
        \->substitute('\\\\', '\', 'g')
endfunc
```

Thanks for the guidance @lacygoill so we're left with @brammool including that and @NoSuck to test if this also works with his filenames!

**Comment by brammool** [comment]
I'll make a best guess for this and send out the version with recent changes.  Then we can further improve.

**Comment by GitMensch** [comment]
Just tested that - fixed with follow-up code of #9228 - I think that can be closed now.
